File: hs/src/Day8.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TransformListComp #-}

module Day8 (day8) where

import Control.Monad (foldM, join)
import Control.Monad.ST (runST)
import Data.Function (on)
import Data.IntMap qualified as IntMap (elems, empty, insertWith, size)
import Data.List (sortOn, tails)
import Data.List.NonEmpty (nonEmpty)
import Data.List.NonEmpty qualified as NE (head, zipWith)
import Data.Maybe (mapMaybe)
import Data.Ord (Down (Down))
import Data.Text (Text)
import Data.Text qualified as T (lines, splitOn, unpack)
import Data.Vector.Unboxed.Mutable qualified as MV (generate, read, write)

day8 :: Int -> Text -> (Int, Maybe Int)
day8 n input = runST $ do
  mapping <- MV.generate size id
  let lookupId i = do
        j <- MV.read mapping i
        if i == j
          then pure i
          else do
            k <- lookupId j
            k <$ MV.write mapping i k
      go part1 components m ((i, j) : edges) = do
        a <- lookupId i
        b <- lookupId j
        components' <-
          if a /= b
            then components - 1 <$ MV.write mapping (max a b) (min a b)
            else pure components
        part1' <-
          if m + 1 /= n
            then pure part1
            else do
              let add counts k = do
                    v <- lookupId k
                    pure $ IntMap.insertWith (+) v 1 counts
              counts <- foldM add IntMap.empty [0 .. size - 1]
              pure $ if IntMap.size counts < 3 then 0 else product . take 3 . sortOn Down $ IntMap.elems counts
        if components' == 1
          then pure (part1', Just $ ((*) `on` NE.head . (nodes !!)) i j)
          else go part1' components' (m + 1) edges
      go part1 _ _ _ = pure (part1, Nothing)
  go 0 size 0 $
    [ (i, j)
    | (i, a) : rest <- tails $ zip [0 ..] nodes,
      (j, b) <- rest,
      let d = sum $ join (*) <$> NE.zipWith (-) a b,
      then
        sortOn
      by
        d
    ]
  where
    nodes = mapMaybe (nonEmpty . map (read . T.unpack) . T.splitOn ",") (T.lines input)
    size = length nodes

File: rs/src/day8.rs
use std::cmp::{Reverse, max, min};
use std::collections::BinaryHeap;

pub fn solve(input: &str, n: usize) -> (usize, Option<isize>) {
    let nodes = input
        .lines()
        .filter_map(|line| {
            line.split(',')
                .map(|s| s.parse::<isize>())
                .collect::<Result<_, _>>()
                .ok()
        })
        .collect::<Vec<Vec<isize>>>();
    let mut edges = nodes
        .iter()
        .enumerate()
        .flat_map(|(i, a)| {
            nodes.iter().enumerate().skip(i + 1).map(move |(j, b)| {
                let d = a
                    .iter()
                    .zip(b.iter())
                    .map(|(a, b)| {
                        let d = a.abs_diff(*b);
                        d * d
                    })
                    .sum::<usize>();
                (Reverse(d), i, j)
            })
        })
        .collect::<BinaryHeap<_>>();
    let mut index = 0;
    let mut part1 = 0;
    let mut components = nodes.len();
    let mut mapping = Mapping((0..nodes.len()).collect());
    while let Some((_, i, j)) = edges.pop() {
        if mapping.merge(i, j) {
            components -= 1;
        }
        if index + 1 == n {
            let mut counts = vec![0; nodes.len()];
            for key in 0..nodes.len() {
                counts[mapping.lookup(key)] += 1;
            }
            counts.retain(|count| *count > 0);
            if counts.len() >= 3 {
                counts.sort_unstable_by_key(|count| Reverse(*count));
                part1 = counts[..3].iter().product();
            }
        }
        if components == 1 {
            return (part1, Some(nodes[i][0] * nodes[j][0]));
        }
        index += 1;
    }
    (part1, None)
}

struct Mapping(Vec<usize>);
impl Mapping {
    fn lookup(&mut self, key: usize) -> usize {
        let mut value = self.0[key];
        if key != value {
            value = self.lookup(value);
            self.0[key] = value;
        }
        value
    }

    fn merge(&mut self, i: usize, j: usize) -> bool {
        let (a, b) = (self.lookup(i), self.lookup(j));
        self.0[max(a, b)] = min(a, b);
        a != b
    }
}
